aten/src/ATen/core/ATenDispatch.h

```diff
@@ -0,0 +1,102 @@
+#pragma once
+
+#include <c10/core/Backend.h>
+#include <unordered_map>
+#include <c10/util/C++17.h>
+#include <memory>
+
+// This dispatch class serves as a replacement for our previous dispatch
+// mechanism, in which all functions were members of a Type class. A derived
+// class existed for each backend (and Variable), and the vtable was used to
+// dispatch to the correct implementation. This class is to be replaced by
+// the c10 dispatcher when it supports all argument and return types.
+// This implementation opts to store implementations in a table of void*.
+
+namespace at {
+
+// ATenOpTable stores the implementations for each backend, in addition to
+// an implementation for variables.
+class CAFFE2_API ATenOpTable {
+ public:
+  ATenOpTable(std::string schema)
+    : schema_(schema) {}
+
+  template<class FuncType>
+  FuncType* getOp(Backend backend, bool is_variable) const {
+    if (is_variable) {
+      return reinterpret_cast<FuncType*>(getVariableOp());
+    }
+    return reinterpret_cast<FuncType*>(getBaseOp(backend));
+  }
+ private:
+  void registerOp(Backend backend, void* fn) {
+    TORCH_CHECK(function_table_[static_cast<int64_t>(backend)] == nullptr,
+        "Attempting to register variable function for schema ", schema_,
+        " and backend ", toString(backend),
+        " but there is already a function registered");
+    function_table_[static_cast<int64_t>(backend)] = fn;
+  }
+
+  void registerVariableOp(void* fn) {
+    TORCH_CHECK(variable_function_ == nullptr,
+        "Attempting to register variable function for schema ", schema_,
+        " but there is already a function registered");
+    variable_function_ = fn;
+  }
+
+  void* getBaseOp(Backend backend) const {
+    if (function_table_[static_cast<int64_t>(backend)] == nullptr) {
+      TORCH_CHECK(function_table_[static_cast<int64_t>(Backend::Undefined)] != nullptr,
+          "No function is registered for schema ", schema_, " on backend ", toString(backend));
```

Comment:
Re error message: not only is no function not registered, but there is no undefined implementation registered either.

Actually, why are we dispatching to `Undefined` at all, at this point? Can't we just error out straight up at this point?